### DISCUSSION TITLE
How does this work

### DISCUSSION THREAD

**Discussion by lorypelli** [vue-vine/vue-vine#167] [Q&A] [answered]
I saw that it uses a tagged template literals tag function `vine`, but I can't find function implementation, only declaration in `.d.ts`. I want to see the code because it seems different behaviour from normal template literals tag functions

**Comment by ShenQingchuan** [comment]
Hi, I'm glad to hear that you're interested in how Vine works, but the issue area is for reporting bugs or requirements, not for asking questions. I'll turn this issue into a discussion and then answer your questions.

**Comment by ShenQingchuan** [comment]
If you have read the documentation of Vue Vine, you'll know that Vine has a compiler to extract the tagged template string from the `.ts` file's AST, and put the string content into `@vue/compiler-dom` to get exactly the same render functions output as Vue SFC.

So If the implementation is all inside the `packages/compiler` folder.

If you have more questions, free to talk with us here.

**Comment by lorypelli** [comment]
> If you have read the documentation of Vue Vine, you'll know that Vine has a compiler to extract the tagged template string from the `.ts` file's AST, and put the string content into `@vue/compiler-dom` to get exactly the same render functions output as Vue SFC.
> 
> So If the implementation is all inside the `packages/compiler` folder.
> 
> If you have more questions, free to talk with us here.

So the `vine` function doesn't have any code, does it?

**Reply by ShenQingchuan** [reply, marked as answer]
Yes, there's no actual implementation of this "function `vine`".

The basic mentality here is to "borrow a form in valid syntax, and handle its AST representation, transform to what you want".

That's a mechanism we called "Macros", it only runs at compile time, and may totally change what you saw in source code.

If you really want to know how this compilation magic works, you can check `src/analyze.ts`, `template/compose.ts`, and `src/transform.ts` under `packages/compiler`.

**Reply by ShenQingchuan** [reply]
Further more, if you're curious about where the `vine` tagged template string function's type is, check out `packages/vue-vine/types/macros.d.ts`, all macros type are here.

We register those type into global and you'll be able to use it everywhere. Of course, this also has a cost, that is, the names of these identifiers are occupied in the global space, so you may need to be careful when naming new variables.

We will gradually add these details to our document, so that everyone can understand better, thank you for your questions and enthusiasm!

**Reply by lorypelli** [reply]
> Further more, if you're curious about where the `vine` tagged template string function's type is, check out `packages/vue-vine/types/macros.d.ts`, all macros type are here.
> 
> We register those type into global and you'll be able to use it everywhere. Of course, this also has a cost, that is, the names of these identifiers are occupied in the global space, so you may need to be careful when naming new variables.
> 
> We will gradually add these details to our document, so that everyone can understand better, thank you for your questions and enthusiasm!

Thanks, but still a question, the template literals sintax that the `vine` function uses it's custom right?

**Reply by lorypelli** [reply]
I mean, for template tag functions it needs to have at least `${}` sintax

**Reply by ShenQingchuan** [reply]
@lorypelli 

No, it's totally OK if you don't use value interpolation i.e. ${...} in JS template string. 

No interpolation means that we call this tag function with no argument.

You can see many this kind of examples in real world use cases, like GraphQL's gql`...`, where we define schema.

This is just a valid JS syntax that we hacked on that.

FYI You can test whether a syntax is valid in any ast explorer.

**Reply by lorypelli** [reply]
> @lorypelli 
> 
> No, it's totally OK if you don't use value interpolation i.e. ${...} in JS template string. 
> 
> No interpolation means that we call this tag function with no argument.
> 
> You can see many this kind of examples in real world use cases, like GraphQL's gql`...`, where we define schema.
> 
> This is just a valid JS syntax that we hacked on that.
> 
> FYI You can test whether a syntax is valid in any ast explorer.

In that case, why not just using normal function, you would have an array with only 1 element, right?

**Reply by ShenQingchuan** [reply]
@lorypelli 

Honestly, I'm not sure what you're doubting about and I'm a little bit confused to your concern in the function form.

As you can see, Vue Vine is designed to borrow the form of declaring component as function, states, effects and methods as function body statements, and we still need somewhere to describe UI. That's why we pick the final return statement, intended to be similar to React.

I think you may not completely understand we are writing PURE Vue template here, i.e. we need a full text content here without any other noises like interpolation expressions. 

What we actually do here is just extracting that string content, put it as the input of `@vue/compiler-dom` in order to compile, and get render functions, that's a necessary step to compose the Vue component object.

The 'vine' function doesn't execute in runtime !!! so I think you may misunderstand it, because it'll be transomed into Vue render functions after Vite plugin processing.

Using tagged template string is just a simple choice, which means we make this syntax in our specific situation ( when tag is 'vine' ) having a different meaning, customized by our own compilation process. ( That's why we call that a "hack" )

You may still think it as what MDN describes:

![Screenshot_2024-11-30-16-06-35-688_com.android.chrome.jpg](https://github.com/user-attachments/assets/27d00bd2-d4b5-4ffc-aedb-212723fdfd6d)

But Vine is not like that in runtime, That component is a function at source code, but an object at runtime.

In our customized use case, you should forget what original usage it was, and treat it as a new syntax in Vue Vine.

We don't change every tagged template string, we don't break all JS/TS syntax, it just works in a function that returns a vine tagged template string.

**Comment by angelhdzmultimedia** [comment]
Loved this discussion! Thanks for asking the question, and for the so kind and useful answers. I was also curious how all this was done. Will play with the source code soon. 

Wishing you all lots of health and peace.
